app/javascript/packages/document-capture-polling/index.ts

```diff
@@ -70,12 +73,15 @@ export class DocumentCapturePolling {
    * @param {boolean} shouldPrompt Whether to bind or unbind page unload behavior.
    */
   bindPromptOnNavigate(shouldPrompt) {
-    window.onbeforeunload = shouldPrompt
-      ? (event) => {
-          event.preventDefault();
-          event.returnValue = '';
-        }
-      : null;
+    const isAlreadyBound = !!this.cleanUpPromptOnNavigate;
+
+    if (shouldPrompt && !isAlreadyBound) {
+      this.cleanUpPromptOnNavigate = promptOnNavigate();
+    } else if (!shouldPrompt && isAlreadyBound) {
+      const cleanUp = this.cleanUpPromptOnNavigate ?? (() => {});
+      this.cleanUpPromptOnNavigate = undefined;
+      cleanUp();
```

Comment:
Kinda unfortunate that TypeScript isn't able to follow the type narrowing through the variable assignment. It could if we avoided the variable, but maybe that sacrifices readability too much?

```suggestion
    if (shouldPrompt && !this.cleanUpPromptOnNavigate) {
      this.cleanUpPromptOnNavigate = promptOnNavigate();
    } else if (!shouldPrompt && this.cleanUpPromptOnNavigate) {
      this.cleanUpPromptOnNavigate();
      this.cleanUpPromptOnNavigate = undefined;
```

Reply:
I prefer the readability of `isAlreadyBound` but yeah I wish TS could deduce that `cleanUpPromptOnNavigate` is there when `isAlreadyBound` is true.